src/passes/LoopInvariantCodeMotion.cpp

```diff
@@ -114,17 +114,20 @@ struct LoopInvariantCodeMotion
       }
       if (interestingToMove(curr)) {
         // Let's see if we can move this out.
-        // Global side effects would prevent this - we might end up
+        // Global state changes would prevent this - we might end up
         // executing them just once.
         // And we must also move across anything not moved out already,
         // so check for issues there too.
         // The rest of the loop's effects matter too, we must also
         // take into account global state like interacting loads and
         // stores.
-        bool unsafeToMove = effects.hasGlobalSideEffects() ||
-                            effectsSoFar.invalidates(effects) ||
-                            (effects.noticesGlobalSideEffects() &&
-                             loopEffects.hasGlobalSideEffects());
+        bool unsafeToMove =
+          effects.writesGlobalState() || effectsSoFar.invalidates(effects) ||
+          (effects.readsGlobalState() && loopEffects.writesGlobalState());
+        // TODO: look into optimizing this with exceptions. for now, disallow
+        if (effects.throws || loopEffects.throws) {
+          unsafeToMove = true;
+        }
```

Comment:
So before it seemed that we allowed the current expression to be moved out of the loop even when the loop threw as long as the current expression does not read any global states, but we don't allow it anymore, right?

What we were doing before seems safe to me, but I guess the difference is pretty marginal anyway, but anyway it is not technically NFC then.. If it's intended this looks fine though.

Reply:
No, I think this is NFC. The old calls to `hasGlobalSideEffects` (for both `effects` and `loopEffects`) would check for `throws`. The refactoring removed that. So to make it NFC, I added them explicitly here.

Reply:
The old code checked for `loopEffects.hasGlobalSideEffects()` only when `effects.noticesGlobalSideEffects()` held; that's what I meant by 
> we allowed the current expression to be moved out of the loop even when the loop threw as long as the current expression did not read any global states

Reply:
Oh, yes, sorry - you're right, I missed that.

I guess I can change that. But would you be ok with leaving it as a TODO for now? It seems low priority, and I do think we can allow full exceptions support there, once we spend the time, anyhow.

Reply:
Yeah that sounds fine. I was mostly just checking.